src/util/netif.cpp

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
+
+#include <util/netif.h>
+
+#include <logging.h>
+#include <netbase.h>
+#include <util/check.h>
+#include <util/sock.h>
+#include <util/syserror.h>
+
+#if defined(__linux__) || defined(__FreeBSD__)
+
+#if defined(__linux__)
+#include <linux/rtnetlink.h>
+#elif defined(__FreeBSD__)
+#include <netlink/netlink.h>
+#include <netlink/netlink_route.h>
+#endif
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+
+    // Create a netlink socket.
+    const int s{socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE)};
+    if (s < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "socket(AF_NETLINK): %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    Sock sock{static_cast<SOCKET>(s)};
+
+    // Send request.
+    struct {
+        nlmsghdr hdr; ///< Request header.
+        rtmsg data; ///< Request data, a "route message".
+        nlattr dst_hdr; ///< One attribute, conveying the route destination address.
+        char dst_data[16]; ///< Route destination address. To query the default route we use 0.0.0.0/0 or [::]/0. For IPv4 the first 4 bytes are used.
+    } request{};
+
+    // Whether to use the first 4 or 16 bytes from request.attr_dst_data.
+    const size_t dst_data_len = network == NET_IPV4 ? 4 : 16;
+
+    request.hdr.nlmsg_type = RTM_GETROUTE;
+    request.hdr.nlmsg_flags = NLM_F_REQUEST;
+#ifdef __linux__
+    request.hdr.nlmsg_flags |= NLM_F_DUMP;
+#endif
+    request.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(rtmsg) + sizeof(nlattr) + dst_data_len);
+    request.hdr.nlmsg_seq = 0; // Sequence number, used to match which reply is to which request. Irrelevant for us because we send just one request.
+    request.data.rtm_family = network == NET_IPV4 ? AF_INET : AF_INET6;
+    request.data.rtm_dst_len = 0; // Prefix length.
+#ifdef __FreeBSD__
+    request.data.rtm_flags = RTM_F_PREFIX;
+#endif
+    request.dst_hdr.nla_type = RTA_DST;
+    request.dst_hdr.nla_len = sizeof(nlattr) + dst_data_len;
+
+    if (sock.Send(&request, sizeof(request), 0) != sizeof(request)) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "send() to netlink socket: %s\n", SysErrorString(errno));
```

Comment:
I realized that in the IPv4 case this would send the trailing 16-4=12 bytes from `request.dst_data[]`. This seems harmless, but better send exactly what's needed:

```suggestion
    if (sock.Send(&request, request.hdr.nlmsg_len, 0) != request.hdr.nlmsg_len) {
        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "send() to netlink socket: %s\n", SysErrorString(errno));
```

Also, I am not sure if we should worry about partial writes with netlink sockets, maybe the send can be interrupted? There is a convenience method `Sock::SendComplete()`:

```cc
    try {
        CThreadInterrupt intr;
        sock.SendComplete(Span{reinterpret_cast<const unsigned char*>(&request), request.hdr.nlmsg_len}, 5s, intr);
    } catch (const std::runtime_error& e) {
        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "writing to netlink socket: %s\n", e.what());
        return std::nullopt;
    }
```

Reply:
It's a datagram (packet socket), so *partial* writes and reads cannot happen. Every packet is interpreted as a new one. Using `SendComplete` would be a bug.

Truncated and corrupted packets could happen in the case of UDP, but as NETLINK is a communication interface with the kernel, that would be rare. No retries are needed. Would still want to detect it and error out gracefully, though, for robustness.